File: 03 - OOP Fundamentals/Abstraction/C++ Code/abstraction.cpp
// =============================================================================
// ABSTRACTION — exposing *what* an object can do (its essential interface) while
// hiding *how* it does it. Callers program to the abstraction, not the details.
//
// Domain: a checkout that accepts different payment methods. The checkout knows
// only "a PaymentMethod can pay(amount)". It has no idea about card networks,
// UPI handshakes, or wallet balances — those details live behind the interface.
//
// Build & run:
//   g++ -std=c++14 abstraction.cpp -o abstraction
//   ./abstraction
// =============================================================================
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// The abstraction: the essential capability, with ZERO implementation detail.
class PaymentMethod {
public:
    virtual bool pay(double amount) = 0;   // pure virtual => abstract: the WHAT, not the HOW
    virtual string name() = 0;
    virtual ~PaymentMethod() {}
};

// Each concrete type hides its own messy "how" behind the same clean "what".
class CreditCard : public PaymentMethod {
private:
    string last4;
public:
    CreditCard(string last4) {
        this->last4 = last4;
    }
    bool pay(double amount) {
        // (hidden detail) tokenize, contact the card network, handle 3-D Secure...
        cout << "  [CreditCard ****" << this->last4 << "] authorizing Rs " << amount << endl;
        return true;
    }
    string name() { return "Credit Card"; }
};

class UpiPayment : public PaymentMethod {
private:
    string vpa;
public:
    UpiPayment(string vpa) {
        this->vpa = vpa;
    }
    bool pay(double amount) {
        // (hidden detail) build a collect request, poll the PSP, await the PIN...
        cout << "  [UPI " << this->vpa << "] collecting Rs " << amount << endl;
        return true;
    }
    string name() { return "UPI"; }
};

// The client depends ONLY on the abstraction. Add a Wallet tomorrow and this
// function never changes — that is the payoff of abstraction.
void checkout(PaymentMethod* method, double amount) {
    cout << "Paying via " << method->name() << "..." << endl;
    if (method->pay(amount)) {
        cout << "  -> success" << endl;
    }
}

int main() {
    vector<PaymentMethod*> methods;
    methods.push_back(new CreditCard("4242"));
    methods.push_back(new UpiPayment("aarav@okbank"));

    for (int i = 0; i < (int)methods.size(); i++) {
        checkout(methods[i], 799.0);
    }

    for (int i = 0; i < (int)methods.size(); i++) {
        delete methods[i];
    }
    return 0;
}

File: 03 - OOP Fundamentals/Abstraction/Java Code/Abstraction.java
// =============================================================================
// ABSTRACTION (Java twin of abstraction.cpp) — the caller depends only on the
// PaymentMethod interface (the WHAT); each implementation hides its own HOW.
//
// Build & run:
//   javac Abstraction.java
//   java Abstraction
// =============================================================================
import java.util.ArrayList;
import java.util.List;

// The abstraction: WHAT you can do, not HOW.
interface PaymentMethod {
    boolean pay(double amount);
    String name();
}

class CreditCard implements PaymentMethod {
    private String last4;
    CreditCard(String last4) { this.last4 = last4; }
    public boolean pay(double amount) {
        // (hidden detail) tokenize, contact the card network, handle 3-D Secure...
        System.out.println("  [CreditCard ****" + this.last4 + "] authorizing Rs " + amount);
        return true;
    }
    public String name() { return "Credit Card"; }
}

class UpiPayment implements PaymentMethod {
    private String vpa;
    UpiPayment(String vpa) { this.vpa = vpa; }
    public boolean pay(double amount) {
        // (hidden detail) build a collect request, poll the PSP, await the PIN...
        System.out.println("  [UPI " + this.vpa + "] collecting Rs " + amount);
        return true;
    }
    public String name() { return "UPI"; }
}

public class Abstraction {
    // The client depends ONLY on the abstraction. Add a Wallet tomorrow and this
    // method never changes.
    static void checkout(PaymentMethod method, double amount) {
        System.out.println("Paying via " + method.name() + "...");
        if (method.pay(amount)) {
            System.out.println("  -> success");
        }
    }

    public static void main(String[] args) {
        List<PaymentMethod> methods = new ArrayList<PaymentMethod>();
        methods.add(new CreditCard("4242"));
        methods.add(new UpiPayment("aarav@okbank"));

        for (PaymentMethod m : methods) {
            checkout(m, 799.0);
        }
    }
}
